### ISSUE TITLE
RedundantModifier mistakenly flags public constructor in inner class of an interface

### ISSUE THREAD

**Issue by cesar1000** [checkstyle/checkstyle#3115]
In the following snippet, RedundantModifier flags SingleUser is not requiring a public constructor, in spite of the class being public in ModuleItem.

```
public interface ModuleItem {
    abstract class BaseModuleItem implements ModuleItem {
        @NotNull private final Module mModule;

        protected BaseModuleItem(@NotNull Module module) {
            mModule = module;
        }
    }

    @Immutable
    class SingleUser extends BaseModuleItem {
        public SingleUser(@NotNull Module module) {
            super(module);
        }
    }
}
```

**Comment by romani** [comment]
please report issue as described here - http://checkstyle.sourceforge.net/report_issue.html#How_to_report_a_bug

**Comment by cesar1000** [comment]
sure :)

```
[tw-mbp-cpuerta checkstyle]$ cat Test.java
interface TestInterface {
  class TestClass {
    // This constructor needs to be public.
    public TestClass() {
    }
  }
}[tw-mbp-cpuerta checkstyle]$ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="RedundantModifier"/>
    </module>
</module>[tw-mbp-cpuerta checkstyle]$ java -jar checkstyle-6.17-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /Users/cpuerta/temp/checkstyle/Test.java:4:5: Redundant 'public' modifier. [RedundantModifier]
Audit done.
Checkstyle ends with 1 errors.
```

My expectation is that the constructor is allowed to be public in this case.

**Comment by romani** [comment]
Can you print to us javac error when public modifier is removed or explain why you think it is required?

All details will help.

**Comment by cesar1000** [comment]
Mmm I don't seem to be able to repro :/. I'll keep an eye out and re-open if it reoccurs!

**Comment by kdubb** [comment]
This is indeed an issue...

Given this definition:
```java
package my.annotations;

import javax.enterprise.util.AnnotationLiteral;

public @interface MyAnn {
  class Literal extends AnnotationLiteral<MyAnn> implements MyAnn {
    Literal() {}
  }
}
```

This code will fail to compile (Java 8/9) with an access issue:
```java
package my.code;

import my.annotations.MyAnn.Literal;

public class MyCode {
  public static void main(String[] args) {
    new MyAnn.Literal();  // Fails because Literal constructor is package-private
  }
}
```

If the `public` modifier is added to the `Literal` constructor, it compiles properly.

@cesar1000  Can we re-open this?

**Comment by kdubb** [comment]
I guess I forgot to mention that `RedundantModifier` claims that in the compiling version:
```java
 class Literal extends AnnotationLiteral<MyAnn> implements MyAnn {
    public Literal() {}
  }
```

The `public` modifier is considered redundant.

**Comment by rnveach** [comment]
How is this any different then https://github.com/checkstyle/checkstyle/issues/2508 ?
The check can not know what is needed to make code work from another class. It can only look at the single file it is currently working with.

**Comment by kdubb** [comment]
I believe it's different because the class _**is**_ public due to where it's defined (inside an interface or annotation). So a public class should allow a public constructor without being redundant.

#2508 seems to be about non-public classes with needed constructors of other protection levels.

**Comment by rnveach** [comment]
They are both the same as they need a public constructor defined by some other code in another class/file.
#2508 deals with serialization routine, while this deals with compile time of 2 different files.

> the class is public due to where it's defined (inside an interface or annotation).

Violation on constructor only appears when the class' `public` modifier is violated on and removed for being redundant.

````
public @interface MyAnn {
  public class Literal { // violation
    public Literal() {} // no violation
````

Even though the class is forced to be `public` the constructor is not when the modifier is taken away. `public` must be declared if the constructor is to be called from outside the file.

@romani I agree the violation shouldn't appear in this situation, for the direct inner class of an interface or annotation. It is definitely not redundant as it directly declares what the access level of the constructor is.
If you agree, please re-open issue.

**Comment by romani** [comment]
@kdubb ,
 Thanks for bringing it back to our attention. You looks like right.

Please open new issue with properly formatted description, files, configuration.

This issue should stay closed. Nobody likes to read long discussions and incomplete steps to reproduce.
